### ISSUE TITLE
fsdp_tp: replace forward_backward closure holders with a structured result

### ISSUE THREAD

**Issue by saforem2** [saforem2/ezpz#212]
Follow-up from review of #207 ([thread](https://github.com/saforem2/ezpz/pull/207#discussion_r0)).

The training loop shuttles state out of `_tinker_step.forward_backward`
through three closure-mutable holders plus a callback:

```python
_t1_holder: list[float] = []
_step_probe: dict[str, object] = {}
_seq_holder: list[int] = []

def _at_forward_done(pred, masked_labels, local_seq_len, ...):
    ...
_fb = _tinker_step.forward_backward(train_state, batch,
                                    on_forward_done=_at_forward_done)
t1 = _t1_holder[0] if _t1_holder else perf_counter()
```

Sourcery's suggestion: have `forward_backward` return a
`ForwardBackwardResult` carrying `t1_barrier` and the forward-done
metrics, so the loop reads `_fb.t1_barrier` / `_fb.metrics` and the
holders disappear.

Not done in #207 because `_at_forward_done` does more than pass state:

- it draws the `t1` barrier at a specific point, so `train/dtf` and
  `train/dtb` keep their current meanings;
- it runs **first-step-only** debug probes, one of which allocates a
  full-logits tensor (the reason `EZPZ_TRACK_LOGITS` exists at all).

Moving that inside `forward_backward` means either running it for every
caller or pushing the same gating down into the step module. Worth
doing, but it reshapes the metrics path — and #207's correctness gate is
a loss-curve comparison against a measured run-to-run noise band, which
a metrics refactor would muddy.

Keep the callback for the Tinker API; the change is about where the
*data* flows.
